src/main/java/edu/harvard/iq/dataverse/authorization/ReverseProxyFilter.java

```diff
@@ -0,0 +1,114 @@
+package edu.harvard.iq.dataverse.authorization;
+
+import edu.harvard.iq.dataverse.util.SystemConfig;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.inject.Inject;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebFilter;
+import javax.servlet.http.HttpFilter;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+@WebFilter( filterName = "reverse-proxy", urlPatterns = "/*")
+public class ReverseProxyFilter extends HttpFilter {
+    
+    @Inject
+    SystemConfig systemConfig;
+    
+    String headerField;
+    
+    private static final Logger logger = Logger.getLogger(ReverseProxyFilter.class.getName());
+    private static final String X_FORWARDED_FOR = "X-Forwarded-For"; // de-facto standard
+    private static final String RFC7239_FORWARDED = "Forwarded";
+    private static HashSet<String> ALLOWED_HEADERS = new HashSet<String>(Arrays.asList(
+        X_FORWARDED_FOR,
+        RFC7239_FORWARDED,
+        "Proxy-Client-IP",
+        "WL-Proxy-Client-IP",
+        "HTTP_X_FORWARDED_FOR",
+        "HTTP_X_FORWARDED",
+        "HTTP_X_CLUSTER_CLIENT_IP",
+        "HTTP_CLIENT_IP",
+        "HTTP_FORWARDED_FOR",
+        "HTTP_FORWARDED",
+        "HTTP_VIA",
+        "REMOTE_ADDR" ));
+    
+    @Override
+    protected void doFilter(HttpServletRequest req, HttpServletResponse res, FilterChain chain) throws IOException, ServletException {
+        HttpServletRequest proxy = req;
+        
+        // process headers and try to find an ip if a header field is present (=configured)
+        if (headerField != null) {
+            String proxiedIp;
+            
+            // get header(s) matching the field and process if not empty (= not present in request)
+            List<String> headers = Collections.list(req.getHeaders(headerField));
+            logger.log(Level.FINER, "Header(s) \""+headerField+"\": "+StringUtils.join(", "));
+            
+            if ( ! headers.isEmpty()) {
+                // special parsing for X-Forwarded-For de-facto standard
+                if (headerField.equals(X_FORWARDED_FOR)) {
+                    proxiedIp = parseXForwardedFor(headers);
+                }
+                // special parsing for RFC 7239 "Forwarded" syntax
+                else if (headerField.equals(RFC7239_FORWARDED)) {
+                    proxiedIp = parseRFC7239Forwarded(headers);
+                }
+                // otherwise get last header (if multiple), which should have been added by the proxy the client hits
```

Comment:
I'm not so sure about this. Copy-cat from #6973. Shouldn't it be the first header that is added by the first proxy? I have never seen this in real life anyway... :sweat_smile:

Reply:
It depends on what you're supporting. The global order is that left to right, first to last goes from closer to the user to closer to the server. So taking the first value, when there's no spoofing, should be the client's IP. That would be what you want for MDC, but not what you want for IP groups, where the fact that you're coming through a campus proxy is the key information. With spoofing, a client can add the first/leftmost value, so taking that will get you spoofed. The only way to assure the first/left-most value is good is if you explicitly can trust all the intermediate proxies. 

For the implementation in 6973, to avoid having to come up with a trusted white list of intermediates, we chose to only support the case of at most a single proxy beyond the LB. In those cases, the last/rightmost value is either the users' IP reported by the LB - good for MDC and IP groups, or it is the proxy IP - good for IP groups and ~OK for MDC (hides the true location, but you'd get the geo-location of the proxy, e.g. the campus for most folks.

I think the 'right' way to do this is with a whitelist, reporting the first/left-most trusted value as the IP address and returning the full list of proxy IPs to process for evaluating IP groups. Given the amount of work for that, and the likelihood that very few places will have long proxy chains - (do we know of anyone with more than an LB and a campus proxy?) we chose the strategy describe here.

Reply:
I guess a meta question here - given that a filter is independent of Dataverse, does an open source one already exist that we can use? I saw one that just grabs the first value, which has the issue discussed above, but perhaps there is one that handles a whitelist of trusted proxies, etc.?